ql/src/java/org/apache/hadoop/hive/ql/ValidTxnManager.java

```diff
@@ -183,8 +183,21 @@ ValidTxnWriteIdList recordValidWriteIds() throws LockException {
   }
 
   private ValidTxnWriteIdList getTxnWriteIds(String txnString) throws LockException {
-    List<String> txnTables = getTransactionalTables(getTables(true, true));
-    ValidTxnWriteIdList txnWriteIds = null;
+
+  List<String> txnTables = getTransactionalTables(getTables(true, true));
+  ValidTxnWriteIdList txnWriteIds = null;
+
+   // If we have collected all required table writeid (in SemanticAnalyzer), skip fetch again
+   if (driverContext.getConf().get(ValidTxnWriteIdList.VALID_TABLES_WRITEIDS_KEY) != null) {
+      txnWriteIds = new ValidTxnWriteIdList(driverContext.getConf().get(ValidTxnWriteIdList.VALID_TABLES_WRITEIDS_KEY));
+      for (String txnTable : txnTables) {
+        if (txnWriteIds.getTableValidWriteIdList(txnTable) == null) {
+          txnWriteIds = null;
+          break;
+        }
+      }
+    }
```

Comment:
If we keep it this place, I would move it inside the else for getCompactionWriteIds, since we do not have DDLs in compaction anyway...

Reply:
Makes sense, done.